### PR TITLE
ARROW-3242: [C++] Make CpuInfo a singleton, use coarser-grained dispatch to SSE4 in Parquet dictionary encoding

### PR DESCRIPTION
**Pull request by wesm**
Rather than having a bunch of static state in a header file (which is not a great pattern anyway), this makes `arrow::CpuInfo` a singleton. I added support for vector-level dispatch to SSE4 hashing in the Parquet dictionary encoder.

This solves ARROW-3241 for me

NB. This is precisely the kind of change that is now radically simpler after the monorepo merge

### CODE REVIEW

**Comment by wesm** [comment]
cc @majetideepak @xhochy @pitrou

File: cpp/CMakeLists.txt [hunk 1]
```diff
@@ -690,7 +690,9 @@ if (ARROW_JEMALLOC)
   add_definitions(-DARROW_JEMALLOC)
   add_definitions(-DARROW_JEMALLOC_INCLUDE_DIR=${JEMALLOC_INCLUDE_DIR})
 
-  if (CMAKE_COMPILER_IS_GNUCXX AND PTHREAD_LIBRARY)
+  if ((CMAKE_COMPILER_IS_GNUCXX OR
+        (CMAKE_CXX_COMPILER_ID MATCHES "Clang"))
+      AND PTHREAD_LIBRARY)
```
Comment (wesm) [inline cpp/CMakeLists.txt]: I found that linking pthread is also necessary for me when using clang-6 on Linux with Ubuntu 14.04

File: cpp/src/arrow/builder.cc [hunk 2]
```diff
@@ -33,12 +33,17 @@
...
 #include "arrow/util/decimal.h"
 #include "arrow/util/hash-util.h"
 #include "arrow/util/hash.h"
 #include "arrow/util/logging.h"
 
+#ifdef ARROW_USE_SSE
+#define SSE4_FLAG true
+#else
+#define SSE4_FLAG false
+#endif
```
Comment (wesm) [inline cpp/src/arrow/builder.cc]: We'll have to do more work if we want better dynamic dispatch to sse4.2 in here

**Comment by wesm** [comment]
cc also @pcmoritz -- this fixes the Plasma test core dump I'd been having

File: cpp/src/parquet/column_writer.cc [hunk 3]
```diff
@@ -32,6 +33,8 @@
 #include "parquet/thrift.h"
 #include "parquet/util/memory.h"
 
+using arrow::CpuInfo;
```
Comment (majetideepak) [inline cpp/src/parquet/column_writer.cc]: Why is this needed here?

**Review by majetideepak — APPROVED** [review]
Apart from a minor clarification, +1 LGTM.

File: cpp/src/parquet/column_writer.cc (hunk 3 above)
Comment (wesm) [reply cpp/src/parquet/column_writer.cc]: Refactor artifact, will remove
